pkg/nodeagent/controller/operatingsystemconfig/changes.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2023 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operatingsystemconfig
+
+import (
+	"errors"
+	"fmt"
+	"slices"
+
+	"github.com/spf13/afero"
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/utils"
+)
+
+var decoder runtime.Decoder
+
+func init() {
+	scheme := runtime.NewScheme()
+	utilruntime.Must(extensionsv1alpha1.AddToScheme(scheme))
+	decoder = serializer.NewCodecFactory(scheme).UniversalDeserializer()
+}
+
+func extractOSCFromSecret(secret *corev1.Secret) (*extensionsv1alpha1.OperatingSystemConfig, []byte, string, error) {
+	oscRaw, ok := secret.Data[dataKeyOperatingSystemConfig]
+	if !ok {
+		return nil, nil, "", fmt.Errorf("no %s key found in OSC secret", dataKeyOperatingSystemConfig)
+	}
+
+	osc := &extensionsv1alpha1.OperatingSystemConfig{}
+	if err := runtime.DecodeInto(decoder, oscRaw, osc); err != nil {
+		return nil, nil, "", fmt.Errorf("unable to decode OSC from secret data key %s: %w", dataKeyOperatingSystemConfig, err)
+	}
+
+	return osc, oscRaw, utils.ComputeSHA256Hex(oscRaw), nil
+}
+
+type operatingSystemConfigChanges struct {
+	units units
+	files files
+}
+
+type units struct {
+	changed []changedUnit
+	deleted []extensionsv1alpha1.Unit
+}
+
+type changedUnit struct {
+	extensionsv1alpha1.Unit
+	dropIns dropIns
+}
+
+type dropIns struct {
+	changed []extensionsv1alpha1.DropIn
+	deleted []extensionsv1alpha1.DropIn
+}
+
+type files struct {
+	changed []extensionsv1alpha1.File
+	deleted []extensionsv1alpha1.File
+}
+
+func computeOperatingSystemConfigChanges(fs afero.Fs, newOSC *extensionsv1alpha1.OperatingSystemConfig) (*operatingSystemConfigChanges, error) {
```

Comment:
```suggestion
func computeOperatingSystemConfigChanges(fs afero.Afero, newOSC *extensionsv1alpha1.OperatingSystemConfig) (*operatingSystemConfigChanges, error) {
```
Same as for token controller. Using `afero.Afero` would shorten the ReadFile call in this function a bit.